### PR TITLE
feat: refactor runner code into multiple runtimes to simplify the addition of new ones

### PR DESCRIPTION
**Pull request by Angelmmiguel**
The `Runner` struct had logic to process requests using different runtimes. Every of them requires a specific configuration for the source code and the WASI environment. Since we only supported two inputs (`javascript` and `wasm`), conditionals were simple and the code was maneagable. However, our goal is to introduce more languages in the future so this specific architecture doesn't fit these requires.

Here, I refactored the `Runner` code into a separate `Worker` module. This module defines both the main `Worker` class and the N runtimes we support. Adding a new runner only requires to add a new struct that `impl` the `Runtime` trait. Then, you only need to configure it in the `init_runtime` function (_pending to implement_). 

Apart from that, I introduced another major change. Now, interpreted languages will mount their source code into the module. Other language requires to mount `lib` folders to work, so it feels natural to use the same approach to all of them. Note that in `Wasmtime` we can preopen folders, but not files. It is an important statement as a folder may contain multiple source files and we want the runtime to have access only to a single one.

For this, I added a new `.wws` folder. This temporary folder contains folders with the source code that it's required for a certain worker. To avoid collisions, every language will have its own scope and every source file a folder with a name based on its context. The folder name is based on the [blake3](https://crates.io/crates/blake3) hash of the file content. I chose this implementation based on the performance metrics they provide.

```
.wws
| - js(lang prefix)
    | - blake3(source-file)
        | - index.(extension)
| - ...
```

That approach opens new possibilities as we can add `glue` code in those folders and even third-party libraries. It will also provides transparency as users will be able to see the code that it's run and even debug it.

## Pending

- [x] Implements the `init_runtime` function to get a `<dyn Runtime + Send + Sync>` instance based on the file extension

### CODE REVIEW

**Review by ereslibre — APPROVED** [review]
Great refactor @Angelmmiguel! LGTM, minor comments from my side.

**Comment by ereslibre** [comment]
Also minor typo in commit title `simlpify`

File: src/workers/runtimes/javascript.rs [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+use super::temp_utils::TempUtils;
+use crate::workers::runtime::Runtime;
+use anyhow::Result;
+use std::{fs, path::PathBuf};
+use wasmtime_wasi::Dir;
+use wasmtime_wasi::WasiCtxBuilder;
+
+static JS_ENGINE_WASM: &[u8] =
+    include_bytes!("../../../kits/javascript/wasm-workers-quick-js-engine.wasm");
```
[the remark this answers is not in the document]
Comment (Angelmmiguel) [reply src/workers/runtimes/javascript.rs:12]: 😄

File: src/workers/runtimes/temp_utils.rs [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+/// Struct to deal with temporary files as part of the
+/// Runtime process. Some runtimes may require to write
+/// temporary files on a folder. Note that Wasm VMs require
+/// to mount a folder, not a file. To keep workers isolated
+/// between others, we will prepare specific folders with the
+/// source code only.
+pub struct TempUtils {
+    /// The folder inside the main TMP folder.
+    pub folder: String,
+}
```
[the remark this answers is not in the document]
Comment (Angelmmiguel) [reply src/workers/runtimes/temp_utils.rs:23]: Good point. It's not clear the purpose of this struct. I will rename it to `Data`

File: src/workers/runtime.rs [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+/// Define the behavior a Runtime must have. This includes methods
+/// to initialize the environment for the given runtime as well as
+/// the Wasi Context to process the request.
+pub trait Runtime {
+    /// Prepare the environment. This method
+    /// initializes folders and create files if required
+    fn prepare(&self) -> Result<()> {
+        Ok(())
+    }
```
[the remark this answers is not in the document]
Comment (Angelmmiguel) [reply src/workers/runtime.rs:19]: Done! 😄

File: src/workers/runtime.rs [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+        Ok(builder)
+    }
+
+    /// Check if the given path can be managed by this runtime.
+    // fn can_manage(path: &Path) -> bool;
+
+    /// Returns a reference to the Wasm module that should
+    /// run this worker. It can be a custom (native) or a
+    /// shared module (others).
+    fn module_bytes(&self) -> Result<Vec<u8>>;
```
[the remark this answers is not in the document]
Comment (Angelmmiguel) [reply src/workers/runtime.rs]: Yes, I had to change the implementation in the end hehe

File: src/workers/runtime.rs [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+    /// Append the required properties to the given
+    /// WASI context builder. This allow runtimes to mount
+    /// specific lib folders, source code and adding
+    /// environment variables.
+    fn prepare_wasi_ctx(&self, builder: WasiCtxBuilder) -> Result<WasiCtxBuilder> {
+        Ok(builder)
+    }
+
+    /// Check if the given path can be managed by this runtime.
+    // fn can_manage(path: &Path) -> bool;
```
[the remark this answers is not in the document]
Comment (Angelmmiguel) [reply src/workers/runtime.rs]: Done! I just removed it 😄

**Pull request merged**